node/pkg/query/request.go

```diff
@@ -0,0 +1,470 @@
+	ChainId vaa.ChainID
+
+	// Query is the chain specific query data.
+	Query ChainSpecificQuery
+}
+
+// ChainSpecificQuery is the interface that must be implemented by a chain specific query.
+type ChainSpecificQuery interface {
+	Type() ChainSpecificQueryType
+	Marshal() ([]byte, error)
+	Unmarshal(data []byte) error
+	UnmarshalFromReader(reader *bytes.Reader) error
+	Validate() error
+}
+
+// ChainSpecificQueryType is used to interpret the data in a per chain query request.
+type ChainSpecificQueryType uint8
+
+// EthCallQueryRequestType is the type of an EVM eth_call query request.
+const EthCallQueryRequestType ChainSpecificQueryType = 1
+
+// EthCallQueryRequest implements ChainSpecificQuery for an EVM eth_call query request.
+type EthCallQueryRequest struct {
+	// BlockId identifies the block to be queried. It mus be a hex string starting with 0x. It may be a block number or a block hash.
+	BlockId string
+
+	// CallData is an array of specific queries to be performed on the specified block, in a single RPC call.
+	CallData []*EthCallData
+}
+
+// EthCallData specifies the parameters to a single EVM eth_call request.
+type EthCallData struct {
+	// To specifies the contract address to be queried.
+	To []byte
+
+	// Data is the ABI encoded parameters to the query.
+	Data []byte
+}
+
+const EvmContractAddressLength = 20
+
+// PerChainQueryInternal is an internal representation of a query request that is passed to the watcher.
+type PerChainQueryInternal struct {
+	RequestID  string
+	RequestIdx int
+	Request    *PerChainQueryRequest
+}
+
+// QueryRequestDigest returns the query signing prefix based on the environment.
+func QueryRequestDigest(env common.Environment, b []byte) ethCommon.Hash {
+	// TODO: should this use a different standard of signing messages, like https://eips.ethereum.org/EIPS/eip-712
+	var queryRequestPrefix []byte
+	if env == common.MainNet {
+		queryRequestPrefix = []byte("mainnet_query_request_000000000000|")
+	} else if env == common.TestNet {
+		queryRequestPrefix = []byte("testnet_query_request_000000000000|")
+	} else {
+		queryRequestPrefix = []byte("devnet_query_request_0000000000000|")
+	}
+
+	return ethCrypto.Keccak256Hash(append(queryRequestPrefix, b...))
+}
+
+// PostSignedQueryRequest posts a signed query request to the specified channel.
+func PostSignedQueryRequest(signedQueryReqSendC chan<- *gossipv1.SignedQueryRequest, req *gossipv1.SignedQueryRequest) error {
+	select {
+	case signedQueryReqSendC <- req:
+		return nil
+	default:
+		return common.ErrChanFull
+	}
+}
+
+//
+// Implementation of QueryRequest.
+//
+
+// Marshal serializes the binary representation of a query request.
+// This method calls Validate() and relies on it to range checks lengths, etc.
+func (queryRequest *QueryRequest) Marshal() ([]byte, error) {
+	if err := queryRequest.Validate(); err != nil {
+		return nil, err
+	}
+
+	buf := new(bytes.Buffer)
+
+	vaa.MustWrite(buf, binary.BigEndian, uint8(1))           // version
```

Comment:
Should version be a global const to make version changes easier or this isn't prime time, yet?